File: python_solution/problem2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Aug 23 20:24:32 2024

@author: sharminalina
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Problem1:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        l1Num = self.convertListNodeToNumber(l1)
        l2Num = self.convertListNodeToNumber(l2)

        l12Num = l1Num + l2Num

        result = self.convertNumberToListNode(l12Num)

        return result

    def convertListNodeToNumber(self, node: ListNode) -> int:
        number = 0
        placeholder = 1

        while node is not None:
            number += node.val * placeholder
            placeholder *= 10
            node = node.next
        return number

    def convertNumberToListNode(self, number: int) -> ListNode:
        if number == 0:
            return ListNode(0)
        
        head = None
        current = None

        while number > 0:
            digit = number % 10
            number //= 10

            newNode = ListNode(digit)

            if head is None:
                head = newNode
                current = head
            else:
                current.next = newNode
                current = current.next
        
        return head

def printListNode(node: ListNode):
    values = []
    while node:
        values.append(str(node.val))
        node = node.next
    print(" -> ".join(values))

def main():
    # Input for l1
    l1_input = input("Enter the first linked list (comma-separated, least significant digit first): ")
    l1_values = list(map(int, l1_input.split(',')))
    l1 = None
    for value in reversed(l1_values):
        l1 = ListNode(value, l1)
    
    # Input for l2
    l2_input = input("Enter the second linked list (comma-separated, least significant digit first): ")
    l2_values = list(map(int, l2_input.split(',')))
    l2 = None
    for value in reversed(l2_values):
        l2 = ListNode(value, l2)

    # Compute the sum
    solution = Problem1()
    result = solution.addTwoNumbers(l1, l2)

    # Print the result
    print("Resulting linked list:")
    printListNode(result)

if __name__ == "__main__":
    main()

File: C#Solutions/problem2.cs
using System;

namespace AllProblems
{
    public class ListNode
    {
        public int val;
        public ListNode next;
        public ListNode(int val = 0, ListNode next = null)
        {
            this.val = val;
            this.next = next;
        }
    }

    public static class Problem2
    {
        public static void Solve()
        {
            Console.WriteLine("Enter the digits for the first number (e.g., 2 4 3 for 342):");
            ListNode l1 = CreateListFromInput(Console.ReadLine());

            Console.WriteLine("Enter the digits for the second number (e.g., 5 6 4 for 465):");
            ListNode l2 = CreateListFromInput(Console.ReadLine());

            ListNode result = AddTwoNumbers(l1, l2);

            Console.WriteLine("Result:");
            PrintListNode(result);
        }

        public static ListNode AddTwoNumbers(ListNode l1, ListNode l2)
        {
            int l1Num = ConvertListNodeToNumber(l1);
            int l2Num = ConvertListNodeToNumber(l2);

            int sum = l1Num + l2Num;

            ListNode result = ConvertNumberToListNode(sum);
            return result;
        }

        static int ConvertListNodeToNumber(ListNode node)
        {
            int number = 0;
            int multiplier = 1;

            while (node != null)
            {
                number += node.val * multiplier;
                multiplier *= 10;
                node = node.next;
            }

            return number;
        }

        static ListNode ConvertNumberToListNode(int number)
        {
            if (number == 0) return new ListNode(0);

            ListNode head = null;
            ListNode current = null;

            while (number > 0)
            {
                int digit = number % 10;
                number /= 10;

                ListNode newNode = new ListNode(digit);

                if (head == null)
                {
                    head = newNode;
                    current = head;
                }
                else
                {
                    current.next = newNode;
                    current = current.next;
                }
            }

            return head;
        }

        static ListNode CreateListFromInput(string input)
        {
            if (string.IsNullOrWhiteSpace(input))
                return null;

            string[] parts = input.Split(' ');
            ListNode head = null;
            ListNode current = null;

            foreach (string part in parts)
            {
                int digit = int.Parse(part);

                ListNode newNode = new ListNode(digit);

                if (head == null)
                {
                    head = newNode;
                    current = head;
                }
                else
                {
                    current.next = newNode;
                    current = current.next;
                }
            }

            return head;
        }

        static void PrintListNode(ListNode node)
        {
            while (node != null)
            {
                Console.Write(node.val);
                if (node.next != null) Console.Write(" -> ");
                node = node.next;
            }
            Console.WriteLine();
        }
    }
}
